File: backend/worker/handlers/replace_audio.py
from __future__ import annotations

from worker.handlers.base import BaseHandler


class ReplaceAudioHandler(BaseHandler):
    async def execute(self, node_config, input_paths, output_path):
        video = input_paths["video"]
        audio = input_paths["audio"]
        loop_if_shorter = bool(node_config.get("loop_if_shorter", True))
        audio_volume = float(node_config.get("audio_volume", 1.0))

        probe = await self.run_ffprobe(video)
        duration = float(probe.get("format", {}).get("duration", 0) or 0)
        if duration <= 0:
            raise RuntimeError("Unable to determine input video duration")

        input_args = ["-i", video]
        if loop_if_shorter:
            input_args.extend(["-stream_loop", "-1", "-i", audio])
        else:
            input_args.extend(["-i", audio])

        filter_chain = f"[1:a]volume={audio_volume}"
        if not loop_if_shorter:
            filter_chain += ",apad"
        filter_chain += "[aout]"

        args = input_args + [
            "-filter_complex", filter_chain,
            "-map", "0:v:0",
            "-map", "[aout]",
            "-c:v", "copy",
            "-c:a", "aac",
            "-t", f"{duration:.3f}",
            output_path,
        ]
        await self.run_ffmpeg(args)
        return {
            "audio_replaced": True,
            "video_duration": duration,
        }

File: internal/worker/handlers/replace_audio.go
package handlers

import (
	"context"
	"errors"
	"fmt"
	"strconv"

	vpffmpeg "github.com/Ctwqk/videoprocess/internal/worker/ffmpeg"
)

type ReplaceAudioHandler struct {
	Runner vpffmpeg.Runner
}

func (h ReplaceAudioHandler) NodeType() string {
	return "replace_audio"
}

func ReplaceAudioArgs(videoPath string, audioPath string, outputPath string, config map[string]any, videoProbe probeSummary) []string {
	loopIfShorter := truthyValue(config["loop_if_shorter"], true)
	audioVolume := floatValue(config["audio_volume"], 1.0)

	inputArgs := []string{"-i", videoPath}
	if loopIfShorter {
		inputArgs = append(inputArgs, "-stream_loop", "-1", "-i", audioPath)
	} else {
		inputArgs = append(inputArgs, "-i", audioPath)
	}

	filterChain := "[1:a]volume=" + formatFloat(audioVolume)
	if !loopIfShorter {
		filterChain += ",apad"
	}
	filterChain += "[aout]"

	args := append([]string{}, inputArgs...)
	args = append(args,
		"-filter_complex", filterChain,
		"-map", "0:v:0",
		"-map", "[aout]",
		"-c:v", "copy",
		"-c:a", "aac",
		"-t", strconv.FormatFloat(videoProbe.Duration, 'f', 3, 64),
		outputPath,
	)
	return args
}

func (h ReplaceAudioHandler) Execute(ctx context.Context, inputPaths map[string]string, outputPath string, config map[string]any) (map[string]any, error) {
	videoPath := inputPaths["video"]
	if videoPath == "" {
		return nil, errors.New("missing input path on video port")
	}
	audioPath := inputPaths["audio"]
	if audioPath == "" {
		return nil, errors.New("missing input path on audio port")
	}
	videoProbe, err := probePath(ctx, h.Runner, videoPath)
	if err != nil {
		return nil, err
	}
	if videoProbe.Duration <= 0 {
		return nil, fmt.Errorf("unable to determine input video duration")
	}
	if err := runFFmpeg(ctx, h.Runner, ReplaceAudioArgs(videoPath, audioPath, outputPath, config, videoProbe)); err != nil {
		return nil, err
	}
	return map[string]any{"audio_replaced": true, "video_duration": videoProbe.Duration}, nil
}
